tox.ini

```diff
@@ -551,23 +556,21 @@ include = multicast*
 exclude_lines =
     # Have to re-enable the standard pragma
     pragma: no cover
-
     except Exception
     # Don't complain if tests don't hit defensive assertion code:
     raise AssertionError
     raise NotImplementedError
     raise ImportError
     raise unittest.SkipTest
     except unittest.SkipTest as skiperr
-
     # Don't complain if non-runnable code isn't run:
-	if __name__ in u'__main__':
-	if __name__ in '__main__':
+    if __name__ in u'__main__':
+    if __name__ in '__main__':
 
 ignore_errors = True
 
 
 [tool:pytest]
-addopts = --cov --cov-append
+addopts = --cov=  --cov-append
```

Comment:
looks like we have bigger problems at the moment for tox:
read up on the last [tox run in CI](https://github.com/reactive-firewall/multicast/actions/runs/10939898051/job/30371546840?pr=118) and not the issue with multiple py envs (including 3.5 and 3.9, and 3.10, and possibly others) reporting:
>  error: subprocess-exited-with-error
>   
>   × Preparing metadata (pyproject.toml) did not run successfully.
>   │ exit code: 1
>   ╰─> [3 lines of output]
>       error in multicast setup command: 'install_requires' must be a string or list > of strings containing valid project/version requirement specifiers; Expected end or > semicolon (after name and no valid version specifier)
>           See https://github.com/reactive-firewall/multicast/requirements.txt
>               ^
>       [end of output]
>   
>   note: This error originates from a subprocess, and is likely not a problem with pip.
> error: metadata-generation-failed

* fortunately the `tox` tests are not MATs so they are not considered blocking, but they are also not expected-failures, so we should fix those first so we can test the changes to the coverage arguments properly.

* clearly we have a bug/misconfiguration in `pyproject.toml`
While it is set to a string with the requirements file, it seems to expect the contents. Can you refresh my memory about the official documentation of `pyproject`

* read `pyproject.toml` and help me debug

Reply:
🤔 I see a mistake in your reasoning regarding 1: the `readFile` function audits/checks inputs not for a full match to `expected_files = ["""E.md""", """requirements.txt"""]`, rather it checks that `any(aexpected_file in filename for aexpected_file in expected_files)` (and `if not` then `raise ...`) which is to say, if 'E.md' (or 'requirements.txt') is in "README.md" then don't `raise ...`(similarly for "LICENSE.md")
Thus, this can not be the primary cause of the error.

* for this we can assume 2 is done (it is done by another part of the test-code and is not reporting them missing for the failing tox runs - https://github.com/reactive-firewall/multicast/actions/runs/10939898051/job/30371169299?pr=118 for py3.9 passed with only PEP517 warnings as are expected)
* Let's look into idea 3 more
* for this we can also assume 4 is done  (it is done by another part of the test-code and is not reporting them missing for the failing tox runs - [https://github.com/reactive-firewall/multicast/actions/runs/10939898051/job/30371163050?pr=118 MATs](https://github.com/reactive-firewall/multicast/actions/runs/10939898051/job/30371333208?pr=118#step:9:104) passed for py3.9 including the requirements.txt format validation test-case)
* for this you may assume 5 is done and the `setup()` call is fine.
* this leaves us with my theory, unless I'm mistaken, which is the bug is in the 'pyproject.toml' config so let's focus on 'pyproject.toml' and 'tox.ini' and contiune debugging.

* read 'pyproject.toml' again and breakdown the configuration for posterity. Reference supporting documentation (from https://packaging.python.org/en/latest/specifications/declaring-project-metadata/ ) or PEP-518 to evaluate the configuration for compliance and possible improvements.

Reply:
holdup, this leads us back to `setup.py` and the `readFile` for the `install_requires` value as you correctly observed earlier. I may have misunderstood; this is the only possible source of the output: "See https://github.com/reactive-firewall/multicast/..."

:thinking: why would tox behave differently than MATs?

Reply:
* I like the idea to parse requirements.txt:
* we have a unit-test in tests/test_deps.py to validate the format, so read `tests/test_deps.py` and focus on finding the regular expression we use to validate the contents of the 'requirements.txt' file 
* let's keep the 'readFile' function in setup.py as is and instead add another function to use the output of readFile("requirements.txt") to parse the requirements.txt file into the form expected by tox for `install_requires`
* we'll need to also handle the cases where our requirements.txt file has multiple versions (which is not supported by `install_requires` any more than comments are) by only preserving the minimum required constraint in those cases.
* also we can insert the new function into setup.py around [line 88](https://github.com/reactive-firewall/multicast/blob/master/setup.py#L88-L90)
* the changes to the code must also be backwards compatible with all versions of python3 (3.0-3.13)
* this approach should strike a better balance where we can use the correct but complex requirements with tools like pip (via `-r ...` arguments for the requirements) and still transform the complex requirements into simple form expected by `install_requires`

Reply:
I'll go-ahead and refactor setup.py with this and do a local test with tox.

In the mean-time please consider a few ways we can test this new function "parse_requirements_for_install_requires":
* read all the test-code in the `tests/` directory and assume the new testing will need to integrate with our current setup (probably a new test suite)
* read up on the project documentation in docs/CI.md and focus on testing related concepts that may be relevant, if any
* search for any relevant learnings you have and list them in a table for posterity
* assume the entry-point for tests remains the same (you can read `./Makefile` and focus on "test", "test-pytest", and "test-style" to find those)
* try to consider doctests, unittests, and new script/tool and select the most compatible approach with the current project's testing goals (this an optional auxiliary criteria)
*

Reply:
* Testing plan looks good to me, proceed with generating possible implementation(s) based on this plan as discussed.
* I've opened a GH issue #119 to track this topic.
* please add a summary walkthrough of our discussion so-far as a comment to #119 for posterity

---
Regarding the changes to "setup.py", I've discovered no obvious regression (caveat: I only tested MATs, via `make test` and style via `make test-style` locally so far)
However another issue presented once the formatting was fixed:

from the log:
>  Installing build dependencies ... done
>  Getting requirements to build wheel ... done
>  Preparing metadata (pyproject.toml) ... error
>  error: subprocess-exited-with-error
>  
>  × Preparing metadata (pyproject.toml) did not run successfully.
>  │ exit code: 1
>  ╰─> [3 lines of output]
>      error in multicast setup command: 'install_requires' must be a string or list of strings containing valid project/version requirement specifiers; Expected package name at the start of dependency specifier
>           [Errno 2] No such file or directory: './requirements.txt'
>           ^
>       [end of output]
>   
>   note: This error originates from a subprocess, and is likely not a problem with pip.
> error: metadata-generation-failed
> 
> × Encountered error while generating package metadata.
> ╰─> See above for output.

* my theory: this looks like the requirements.txt file is not making it to the tox virtual env

* note the requirements.txt file is present in the repository (the question is how to ensure it is found by tox)
* read all the config files in the project and help me debug

Reply:
Thanks, let's continue:
* please refresh my memory on when the options `skipsdist` and `usedevelop` were added to tox? and where I can find that officially documented.
* please refresh my memory on `MANIFEST.in` and its full syntax, and where I can find that officially documented.
* please refresh my memory on the expected format of `CHANGELOG.md` and its contents, and where I can find that officially documented.
* please refresh my memory on the expected format of `HISTORY.md` and its contents, and where I can find that officially documented.

* please refresh my memory on the impacts of PEP-517 and PEP-518 on the `MANIFEST.in` and various relevant `*.md` files.

Next steps:
* regarding your [earlier comment](https://github.com/reactive-firewall/multicast/pull/118#discussion_r1767628067) about how we could go about testing these new functions, please update and improve the plan to account for the MANIFEST.in file and the "Rebuild the Source Distribution" steps which of-course expands what needs testing (this should be a simple addition, unless I'm mistaken)

Reply:
* ok I've debugged my local copy for the most part, I'll push the fix shortly to trigger CI
* no issue with the added options of `--cov= ` (with extra space was observed)
* please read #119 and add a second supplementary summary walkthrough of our discussion since the last summary, as a comment to https://github.com/reactive-firewall/multicast/issues/119 for posterity

Reply:
looks like it is fixed in PR #121 - Thanks for the help debugging 🙇🏻 

> Fix worked: https://github.com/reactive-firewall/multicast/actions/runs/10952835394/job/30412304622?pr=121 no-longer has the `install_requires` issue 🎉